### PR TITLE
feat: section ordering lint rule

### PR DESCRIPTION
**Pull request by adthrasher**
This pull request adds a new rule to `wdl-lint`.

- **Rule Name**: `SectionOrdering`

This rule ensures that tasks and workflows have their sections in the desired order.

Before submitting this PR, please make sure:

- [x] You have added a few sentences describing the PR here.
- [x] You have added yourself or the appropriate individual as the assignee.
- [x] You have added at least one relevant code reviewer to the PR.
- [x] Your code builds clean without any errors or warnings.
- [x] You have added an entry to the relevant `CHANGELOG.md` (see
      ["keep a changelog"] for more information).
- [x] Your commit messages follow the [conventional commit] style.

Rule specific checks:

- [x] You have added the rule as an entry within `RULES.md`.
- [x] You have added the rule to the `rules()` function in `wdl-lint/src/lib.rs`.
- [x] You have added a test cases in `wdl-lint/tests/lints` that covers every 
      possible diagnostic emitted for the rule within the file where the rule  
      is implemented.
- [x] You have run `wdl-gauntlet --refresh` to ensure that there are no 
      unintended changes to the baseline configuration file (`Gauntlet.toml`).
- [x] You have run `wdl-gauntlet --refresh --arena` to ensure that all of the 
      rules added/removed are now reflected in the baseline configuration file 
      (`Arena.toml`).

### CODE REVIEW

File: wdl-lint/src/rules/section_order.rs [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+        workflow: &WorkflowDefinition,
+    ) {
+        if reason == VisitReason::Exit {
+            return;
+        }
+
+        let mut sections: Vec<usize> = Vec::new();
+
+        if let Some(meta) = workflow.metadata().next() {
+            println!("{:?}", meta);
```
Comment (claymcleod) [inline wdl-lint/src/rules/section_order.rs:128]: This should be removed.

File: wdl-lint/tests/lints/curly-command/source.errors [hunk 2]
```diff
@@ -1,3 +1,11 @@
+note[SectionOrdering]: sections are not in order for task bad
+  ┌─ tests/lints/curly-command/source.wdl:5:6
+  │
+5 │ task bad {
```
Comment (claymcleod) [inline wdl-lint/tests/lints/curly-command/source.errors:4]: For a lint like this, do we want the reported span to encompass the whole body of the task? Or just the name? I haven't kept up with how we have handled this historically.

**Review by claymcleod — APPROVED** [review]
I'm good with this. I'd take a look at my question about which span to report and also removing the `println!`. One last thing to consider is whether we want to report in the fix only the sections that were actually detected rather than a static list of all possible sections. I don't think that's required, but it might be just a tad bit cleaner—if nothing else, for future work.

**Comment by claymcleod** [comment]
Ah, also need to add this to be compatible with #110 when it's merged in.


```
    fn document(&mut self, _: &mut Self::State, reason: VisitReason, _: &Document) {
        if reason == VisitReason::Exit {
            return;
        }

        // Reset the visitor upon document entry
        *self = Default::default();
    }
```

**Comment by peterhuene** [comment]
I'm happy to also rebase my branch if we want to merge this first.

File: wdl-lint/src/rules/section_order.rs [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+        reason: VisitReason,
+        task: &TaskDefinition,
+    ) {
+        if reason == VisitReason::Exit {
+            return;
+        }
+
+        let mut sections: Vec<usize> = Vec::new();
+
+        if let Some(meta) = task.metadata().next() {
```
[the remark this answers is not in the document]
Comment (peterhuene) [reply wdl-lint/src/rules/section_order.rs:81]: Something like:

```rust
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
enum State {
    Start,
    Meta,
    ParameterMeta,
    Input,
    Decl,
    Output
}

let mut state = State::Start;
for item in section.items() {
    match item {
        TaskItem::Metadata(_) if state <= State::Meta {
            state = State::Meta;
        }
        TaskItem::ParameterMetadata(_) if state <= State::ParameterMeta => {
            state = State::ParameterMeta;
        }
        TaskItem::Input(_) if state <= State::Input => {
            state = State::Input;
        }
        TaskItem::Declaration(_) if state <= State::Decl => {
            state = State::Decl;
        }
        TaskItem::Output(_) if state <= State::Output => {
            state = State::Output;
        }
        TaskItem::Command(_) | TaskItem::Runtime(_) => continue,
        _ => {
            state.add(task_section_order(task.name().span(), task.name().as_str()));
            break;
        }
    }
}
```

File: wdl-lint/tests/lints/curly-command/source.errors (hunk 2 above)
Comment (peterhuene) [reply wdl-lint/tests/lints/curly-command/source.errors:4]: I think for this lint we could probably label the task name with "this task contains sections that are out of order" with a secondary label on the span of the keyword for the first out-of-order section we encounter with "this section is out of order".

File: wdl-lint/tests/lints/curly-command/source.errors [hunk 4]
```diff
@@ -1,3 +1,11 @@
+note[SectionOrdering]: sections are not in order for task bad
```
Comment (peterhuene) [inline wdl-lint/tests/lints/curly-command/source.errors:1]: I think we should add an `#@ except` for this lint to the existing files and create a dedicated test for the lint.

**Review by peterhuene — CHANGES_REQUESTED** [review]
This looks really good, just a few comments for your review.

File: wdl-lint/tests/lints/curly-command/source.errors (hunk 2 above)
Comment (adthrasher) [reply wdl-lint/tests/lints/curly-command/source.errors]: Thanks for the suggestion. I didn't realize that you could label two spans (I suppose I could've looked at the docs). Now it highlights the `task` or `workflow` and then adds a secondary highlight to the first offending section that it encounters.

File: wdl-lint/src/rules/section_order.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (adthrasher) [reply wdl-lint/src/rules/section_order.rs]: Thanks @peterhuene! This is an excellent suggestion.

File: wdl-lint/tests/lints/missing-meta/source.wdl [hunk 5]
```diff
@@ -1,3 +1,4 @@
+#@ except: SectionOrdering
```
Comment (peterhuene) [inline wdl-lint/tests/lints/missing-meta/source.wdl:1]:
```suggestion
#@ except: SectionOrdering

```

File: wdl-lint/src/rules/section_order.rs [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+
+/// The identifier for the section ordering rule.
+const ID: &str = "SectionOrdering";
+
+/// Creates a workflow section order diagnostic.
+fn workflow_section_order(span: Span, name: &str, problem_span: Span) -> Diagnostic {
+    Diagnostic::note(format!("sections are not in order for workflow {}", name))
+        .with_rule(ID)
+        .with_highlight(span)
+        .with_label("first problematic section", problem_span)
```
Comment (peterhuene) [inline wdl-lint/src/rules/section_order.rs:29]: For labels, I like to follow what rustc does and use language such as "this" or "here".

```suggestion
        .with_label("this section is out of order", problem_span)
```

File: wdl-lint/src/rules/section_order.rs [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+             `output`",
+        )
+}
+
+/// Creates a task section order diagnostic.
+fn task_section_order(span: Span, name: &str, problem_span: Span) -> Diagnostic {
+    Diagnostic::note(format!("sections are not in order for task {}", name))
+        .with_rule(ID)
+        .with_highlight(span)
+        .with_label("first problematic section", problem_span)
```
Comment (peterhuene) [inline wdl-lint/src/rules/section_order.rs:41]:
```suggestion
        .with_label("this section is out of order", problem_span)
```

File: wdl-lint/src/rules/section_order.rs [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+use crate::Rule;
+use crate::Tag;
+use crate::TagSet;
+
+/// The identifier for the section ordering rule.
+const ID: &str = "SectionOrdering";
+
+/// Creates a workflow section order diagnostic.
+fn workflow_section_order(span: Span, name: &str, problem_span: Span) -> Diagnostic {
+    Diagnostic::note(format!("sections are not in order for workflow {}", name))
```
Comment (peterhuene) [inline wdl-lint/src/rules/section_order.rs:26]:
```suggestion
    Diagnostic::note(format!("sections are not in order for workflow `{name}`"))
```

File: wdl-lint/src/rules/section_order.rs [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+        .with_rule(ID)
+        .with_highlight(span)
```
Comment (peterhuene) [inline wdl-lint/src/rules/section_order.rs:28]: I think it's best to use `with_highlight` only when there's a single span being called attention to.

When there are multiple, it's best to use `with_label` to distinguish them.

```suggestion
        .with_label("this workflow contains sections that are out of order", span)
```

File: wdl-lint/src/rules/section_order.rs [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+        .with_label("first problematic section", problem_span)
+        .with_fix(
+            "order as `meta`, `parameter_meta`, `input`, private declarations/calls/scatters, \
+             `output`",
+        )
+}
+
+/// Creates a task section order diagnostic.
+fn task_section_order(span: Span, name: &str, problem_span: Span) -> Diagnostic {
+    Diagnostic::note(format!("sections are not in order for task {}", name))
```
Comment (peterhuene) [inline wdl-lint/src/rules/section_order.rs:38]:
```suggestion
    Diagnostic::note(format!("sections are not in order for task `{name}`"))
```

File: wdl-lint/src/rules/section_order.rs [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+        .with_rule(ID)
+        .with_highlight(span)
```
Comment (peterhuene) [inline wdl-lint/src/rules/section_order.rs:40]:
```suggestion
        .with_label("this task contains sections that are out of order", span)
```

Comment (peterhuene) [reply wdl-lint/src/rules/section_order.rs:40]: Oh one last thing, let's reorder the labels; the first label should be the first out of order section; that way the line number for any short rendering of the diagnostic points at the section and not the task/workflow.

**Review by peterhuene — APPROVED** [review]
Looks good!

**Pull request merged**